### PR TITLE
bpo-35621: Support running subprocesses in asyncio when loop is executed in non-main thread 

### PR DESCRIPTION
**Pull request by asvetlov**
The resurrection of #13630 
Now without dangling threads in tests :)

https://bugs.python.org/issue35621

### CODE REVIEW

**Comment by asvetlov** [comment]
I tested it with manual adding 1-second timeout in spawned waiter threads to emulate slow test machine.
After making sure that `./python -Werror -m test -v --fail-env-changed test_asyncio` doesn't produce an error I removed the sleep.

File: Lib/asyncio/unix_events.py [hunk 1]
```diff
@@ -1035,11 +1052,6 @@ def __exit__(self, a, b, c):
     def add_child_handler(self, pid, callback, *args):
         assert self._forks, "Must use the context manager"
 
-        if self._loop is None:
```
Comment (asvetlov) [inline Lib/asyncio/unix_events.py:1038]: The check is replaced by `watcher.is_active()` call in subprocess transport

File: Doc/library/asyncio-policy.rst [hunk 2]
```diff
@@ -184,23 +186,64 @@ implementation used by the asyncio event loop:
 
       Note: loop may be ``None``.
 
+   .. method:: is_active()
+
+      Returns ``True`` is the watcher is ready to use, e.g. :py:data:`SIGCHLD`
```
Comment (1st1) [inline Doc/library/asyncio-policy.rst]:
```suggestion
      Return ``True`` if the watcher is ready to use.
```

(Fix a typo; drop the "e.g." and everything after it)

Comment (1st1) [reply Doc/library/asyncio-policy.rst]: "Spawning a subprocess with *inactive* current child watcher raises :exc:`RuntimeError`." on a new line.

File: Doc/library/asyncio-policy.rst [hunk 3]
```diff
@@ -184,23 +186,64 @@ implementation used by the asyncio event loop:
...
+
+   This implementation starts a new waiting thread for every subprocess spawn.
+
+   The watcher can be used by asyncio code executed in non-main thread without any
+   limitation.
+
+   There is no noticeable overhead when handling a big number of children (*O(1)* each
+   time a child terminates) but stating a thread per process is not free.
+
+   asyncio uses this safe implementation by default.
```
Comment (1st1) [inline Doc/library/asyncio-policy.rst]:
```suggestion
   This watcher is used by default.
```

File: Doc/library/asyncio-policy.rst [hunk 4]
```diff
@@ -184,23 +186,64 @@ implementation used by the asyncio event loop:
...
-.. class:: SafeChildWatcher
+.. class:: ThreadedChildWatcher
+
+   This implementation starts a new waiting thread for every subprocess spawn.
+
+   The watcher can be used by asyncio code executed in non-main thread without any
+   limitation.
+
+   There is no noticeable overhead when handling a big number of children (*O(1)* each
+   time a child terminates) but stating a thread per process is not free.
```
Comment (1st1) [inline Doc/library/asyncio-policy.rst]:
```suggestion
   time a child terminates), but stating a thread per process requires extra memory.
```

File: Doc/library/asyncio-policy.rst [hunk 5]
```diff
@@ -184,23 +186,64 @@ implementation used by the asyncio event loop:
...
       This method has to be called to ensure that underlying
       resources are cleaned-up.
 
-.. class:: SafeChildWatcher
+.. class:: ThreadedChildWatcher
+
+   This implementation starts a new waiting thread for every subprocess spawn.
+
+   The watcher can be used by asyncio code executed in non-main thread without any
+   limitation.
```
Comment (1st1) [inline Doc/library/asyncio-policy.rst]: -> "This watcher implementation works reliably even when the asyncio event loop is run in a non-main OS thread".

File: Doc/library/asyncio-policy.rst [hunk 6]
```diff
@@ -184,23 +186,64 @@ implementation used by the asyncio event loop:
...
+
+   asyncio uses this safe implementation by default.
+
+   .. versionadded:: 3.8
+
+.. class:: MultiLoopChildWatcher
 
-   This implementation avoids disrupting other code spawning processes
+   This implementation registers a :py:data:`SIGCHLD` signal handler on
+   instantiation (which may conflict with other code that install own handler for this
```
Comment (1st1) [inline Doc/library/asyncio-policy.rst]:
```suggestion
   instantiation. That can break third-party code that installs a custom handler for `SIGCHLD`.
```

File: Doc/library/asyncio-subprocess.rst [hunk 7]
```diff
@@ -293,18 +293,23 @@ their completion.
...
+On UNIX *child watchers* are used for subprocess finish waiting, see
+:ref:`asyncio-watchers` for more info.
 
-Note that alternative event loop implementations might not share
-the above limitations; please refer to their documentation.
+
+.. versionchanged:: 3.8
+
+   UNIX switched to use :class:`ThreadedChildWatcher` for spawning subprocesses from
+   different threads without any limitation.
```
Comment (1st1) [inline Doc/library/asyncio-subprocess.rst:309]: Please also mention that asyncio will refuse to spawn a subprocess if no child watcher is active.

File: Lib/asyncio/unix_events.py [hunk 8]
```diff
@@ -838,6 +847,15 @@ def close(self):
         """
         raise NotImplementedError()
 
+    def is_active(self):
+        """Watcher status.
```
Comment (1st1) [inline Lib/asyncio/unix_events.py]:
```suggestion
        """Return ``True`` if the watcher is active and is used by the event loop.
```

File: Lib/asyncio/unix_events.py [hunk 9]
```diff
@@ -1101,6 +1113,209 @@ def _do_waitpid_all(self):
                 callback(pid, returncode, *args)
 
 
+class MultiLoopChildWatcher(AbstractChildWatcher):
+    """A watcher that doen't require running loop in main thread to work.
```
Comment (1st1) [inline Lib/asyncio/unix_events.py]:
```suggestion
    """A watcher that doesn't require running loop in the main thread.
```

**Review by 1st1 — APPROVED** [review]
LGTM overall. Thanks for pushing this through, Andrew.

**Comment by asvetlov** [comment]
Thanks for review!

**Pull request merged**